lib/bankster/bic.ex

```diff
@@ -1,23 +1,148 @@
+          location: binary(),
+          branch: binary() | nil
+        }
+
+  @bic_regex ~r/^(?<bank>[a-zA-Z]{4})\s*(?<country>[a-zA-Z]{2})\s*(?<location>[0-9a-zA-Z]{2})\s*(?<branch>[0-9a-zA-Z]{3})?$/
 
-  ## -- Functions
   @doc """
-  Validates a given string whether it's a valid SWIFT BIC
+  Parses a BIC string into its components.
+
+  ## Examples
+    iex> Bankster.Bic.new("INVALIDBIC")
+    :error
+
+    iex> Bankster.Bic.new("AAAABBCC123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+
+    iex> Bankster.Bic.new("AAAABBCC")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: nil}
+
+    iex> Bankster.Bic.new("AAAA BB CC 123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+
+    iex> Bankster.Bic.new("AAAA BB CC")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: nil}
+
+    iex> Bankster.Bic.new("AAAA BB CC     123")
+    %Bankster.Bic{bank: "AAAA", country: "BB", location: "CC", branch: "123"}
+  """
+  @spec new(binary()) :: {:ok, t()} | :error
+  def new(bic) when is_binary(bic) do
+    case Regex.named_captures(@bic_regex, bic) do
+      nil ->
+        :error
+
+      captures ->
+        {
+          :ok,
+          %__MODULE__{
+            bank: captures["bank"],
+            country: captures["country"],
+            location: captures["location"],
+            branch: presence(captures["branch"])
+          }
+        }
+    end
+  end
+
+  def new(_), do: :error
+
+  @doc """
+  Turns a `Bankster.Bic` into a string.
+
+  ## Options
+
+  * `:format` - Specifies the format to use when encoding.
+  * `:case` - Specifies the character case when encoding.
+
+  The values for `:format` can be:
+
+  * `:compact` - No spaces between components. (default)
+  * `:pretty` - Adds spaces between components
+
+  The values for `:case` can be:
+
+  * `:upper` - All characters are made to be upper case. (default)
+  * `:lower` - All characters are made to be lower case.
+  * `:leave` - Leaves the characters in the case they are in.
+  """
+  @spec to_string(t()) :: binary()
+  @spec to_string(t(), [{:format, :pretty | :compact}]) :: binary()
+  def to_string(
+        %{bank: bank, country: country, location: location, branch: branch} = bic,
+        opts \\ []
+      ) do
+    if bank == nil do
+      raise ArgumentError, ":bank is required, got: #{inspect(bic)}"
+    end
+
+    if country == nil do
+      raise ArgumentError, ":country is required, got: #{inspect(bic)}"
+    end
+
+    if location == nil do
+      raise ArgumentError, ":location is required, got: #{inspect(bic)}"
+    end
+
+    pretty? = opts[:format] == :pretty
+
+    string =
+      IO.iodata_to_binary([
+        bank,
+        if(pretty?, do: [~c" "], else: []),
+        country,
+        if(pretty?, do: [~c" "], else: []),
+        location,
+        if(pretty? && branch, do: [~c" "], else: []),
+        if(branch, do: branch, else: [])
+      ])
```

Comment:
I'm not happy with this if I am honest. The alternate implementation I had was:

```elixir
case opts[:format] do
  :pretty ->
    if bic.branch do
      "#{bic.bank} #{bic.country} #{bic.location} #{bic.branch}"
    else
      "#{bic.bank} #{bic.country} #{bic.location}"
    end

  _ ->
    "#{bic.bank}#{bic.country}#{bic.location}#{bic.branch}"
end
```

Reply:
I agree with your opinion, the alternate implementation looks better/clearer to me.